dandi/dandifs.py

```diff
@@ -0,0 +1,434 @@
+"""
+A `fsspec` File System for (remote) DANDI
+"""
+__all__ = ['RemoteDandiFileSystem']
+try:
+    from fsspec.spec import AbstractFileSystem
+    from fsspec.implementations.http import HTTPFileSystem
+    from fsspec.utils import stringify_path, tokenize
+except (ImportError, ModuleNotFoundError):
+    raise ImportError(
+        'fsspec with http support is required by the dandifs submodule. '
+        'Please install with: pip install fsspec[http]')
+import re
+import requests
+from typing import Optional, Union, Tuple
+from urllib.parse import unquote as url_unquote
+from dandi.dandiapi import (
+    DandiAPIClient, RemoteDandiset, RemoteAsset, DandiInstance, NotFoundError)
+from dandi.utils import get_instance
+
+
+class RemoteDandiFileSystem(AbstractFileSystem):
+    """
+    A file system that browses through a remote dandiset.
+
+    Examples
+    --------
+    Load and parse a remote file:
+
+    .. code-block:: python
+
+        from dandi.dandifs import RemoteDandiFileSystem
+        import json
+        fs = RemoteDandiFileSystem()
+        with fs.open('dandi://dandi/000026/rawdata/sub-I38/ses-MRI/anat/'
+                    'sub-I38_ses-MRI-echo-4_flip-4_VFA.json') as f:
+            info = json.load(f)
+
+    The 'dandi://' protocol is registered with fsspec, so the same
+    result can be achived by
+
+    .. code-block:: python
+
+        import fsspec
+        import json
+        with fsspec.open('dandi://dandi/000026/rawdata/sub-I38/ses-MRI/anat/'
+                        'sub-I38_ses-MRI-echo-4_flip-4_VFA.json') as f:
+            info = json.load(f)
+
+    Browse a dataset:
+
+    .. code-block:: python
+
+        from dandi.dandifs import RemoteDandiFileSystem
+        fs = RemoteDandiFileSystem('000026')
+        fs.glob('**/anat/*.json')
+
+    """
+
+    def __init__(
+            self,
+            dandiset: Optional[Union[str, RemoteDandiset]] = None,
+            version: Optional[str] = None,
+            client: Optional[Union[str, DandiInstance, DandiAPIClient]] = None,
+            **http_kwargs):
```

Comment:
What if a user supplies a `RemoteDandiset` with one version, a `version` string with a different version, and a client that doesn't match the `RemoteDandiset`'s client?  Such mismatches should either result in immediate errors or else be prevented in the first place via dedicated constructor classmethods.